File: android/src/main/java/lt/imas/react_native_signal/RNSignalClientModule.java
package lt.imas.react_native_signal;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.whispersystems.libsignal.SignalProtocolAddress;
import org.whispersystems.libsignal.UntrustedIdentityException;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import io.sentry.Sentry;
import lt.imas.react_native_signal.helpers.Base64;
import lt.imas.react_native_signal.signal.LegacyMessage;
import lt.imas.react_native_signal.signal.LogSender;
import lt.imas.react_native_signal.signal.MessageStorage;
import lt.imas.react_native_signal.signal.SignalClient;
import lt.imas.react_native_signal.signal.SignalServer;
import lt.imas.react_native_signal.signal.ProtocolStorage;

import static lt.imas.react_native_signal.signal.PromiseRejectCode.ERR_NATIVE_FAILED;
import static lt.imas.react_native_signal.signal.PromiseRejectCode.ERR_WRONG_CONFIG;
import timber.log.Timber;

public class RNSignalClientModule extends ReactContextBaseJavaModule {
    private LogSender logSender = LogSender.getInstance();

    private SignalClient signalClient;
    private ProtocolStorage protocolStorage;
    private MessageStorage messageStorage;
    private int signalResetVersion = 0;

    private String username;

    public RNSignalClientModule(ReactApplicationContext reactContext) {
        super(reactContext);

        Timber.plant(new Timber.DebugTree());

        String absolutePath = getReactApplicationContext().getFilesDir().getAbsolutePath();
        protocolStorage = new ProtocolStorage(absolutePath);
        messageStorage = new MessageStorage(absolutePath);
    }

    @Override
    public String getName() {
        return "SignalClient";
    }

    @ReactMethod
    public void init(ReadableMap config, Promise promise){
        try {
            if (!config.hasKey("host")
                    || !config.hasKey("username")
                    || !config.hasKey("accessToken")) {
                promise.reject(ERR_WRONG_CONFIG, "Wrong config provided.");
            } else {
                logSender.init(config);

                String accessToken = config.getString("accessToken");
                String host = config.getString("host");
                username = config.getString("username");

                SignalServer signalServer = new SignalServer(host, accessToken, getReactApplicationContext());
                signalClient = new SignalClient(signalServer, protocolStorage, messageStorage, signalResetVersion);

                int existingSignalResetVersion = protocolStorage.getSignalResetVersion();

                // ---
                // check pre key amount in order to reset from a version which was generating large amounts of prekeys
                boolean performSoftReset = protocolStorage.getLastPreKeyIndex() > 300;
                // ---

                if (!performSoftReset && protocolStorage.getLocalUsername().equals(username) && protocolStorage.isLocalRegistered()){
                    signalClient.checkRemotePreKeys(promise);
                } else {
                    protocolStorage.deleteAll();
                    if (performSoftReset) {
                        logSender.sendInfo(String.format("Android Signal soft reset version triggered: %s", signalResetVersion));
                    } else {
                        messageStorage.deleteAll();
                    }
                    registerAccount(promise);
                }
            }
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void registerAccount(final Promise promise){
        try {
            signalClient.registerAccount(username, promise);
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void resetAccount(Promise promise){
        try {
            protocolStorage.deleteAll();
            messageStorage.deleteAll();
            promise.resolve("ok");
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void addContact(ReadableMap params, boolean forceAdd, Promise promise){
        try {
            String username = params.getString("username");
            SignalProtocolAddress address = new SignalProtocolAddress(username, 1);
            if (!protocolStorage.containsSession(address) || forceAdd){
                signalClient.requestPreKeys(username, promise);
            } else {
                promise.resolve("ok");
            }
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void deleteContact(String username, Promise promise){
        try {
            SignalProtocolAddress address = new SignalProtocolAddress(username, 1);
            protocolStorage.removeIdentity(address);
            protocolStorage.deleteSession(address);
            messageStorage.deleteAllContactMessages(username);
            signalClient.deleteContactPendingMessages(username, "*", promise);
            promise.resolve("ok");
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void deleteContactMessages(String username, String tag, Promise promise){
        try {
            messageStorage.deleteContactMessages(username, tag);
            signalClient.deleteContactPendingMessages(username, tag, promise);
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void receiveNewMessagesByContact(String username, String tag, final Promise promise){
        try {
            signalClient.getContactMessages(username, tag, promise, true);
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getMessagesByContact(String username, String tag, final Promise promise){
        try {
            JSONArray messagesJSONA = messageStorage.getContactMessages(username, tag);
            ArrayList<JSONObject> messagesList = new ArrayList<>();
            for (int i = 0; i < messagesJSONA.length(); i++)
                messagesList.add(messagesJSONA.optJSONObject(i));
            Collections.sort(messagesList, new Comparator<JSONObject>() {
                @Override
                public int compare(JSONObject o1, JSONObject o2) {
                    long ts1 = o1.optLong("savedTimestamp");
                    long ts2 = o2.optLong("savedTimestamp");
                    return Long.compare(ts2, ts1);
                }
            });
            promise.resolve(new JSONArray(messagesList).toString());
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getUnreadMessagesCount(String tag, final Promise promise){
        try {
            signalClient.getContactMessages("", tag, promise, false);
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getExistingMessages(String tag, final Promise promise){
        try {
            promise.resolve(messageStorage.getExistingMessages(tag).toString());
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void sendSilentMessageByContact(String tag, ReadableMap params, final Promise promise) {
        try {
            signalClient.sendMessage(params.getString("username"), params.getString("message"), tag, true, promise);
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void sendMessageByContact(String tag, ReadableMap params, final Promise promise) {
        try {
            signalClient.sendMessage(params.getString("username"), params.getString("message"), tag, false, promise);
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setFcmId(String fcmId, final Promise promise) {
        try {
            signalClient.setFcmId(fcmId, promise);
        } catch (Throwable e) {
            logSender.reportError(e);
            promise.reject(e);
        }
    }

    @ReactMethod
    public void prepareApiBody(String tag, ReadableMap params, final Promise promise) {
        try {
            JSONObject request = signalClient.prepareApiBody(params.getString("username"), params.getString("message"), tag, false);
            promise.resolve(request.toString());
        } catch (JSONException
                | IllegalArgumentException
                | UntrustedIdentityException
                | UnsupportedEncodingException e) {
            logSender.reportError(e);
            promise.reject(ERR_NATIVE_FAILED, e.getMessage());
        }
    }

    @ReactMethod
    public void saveSentMessage(String tag, ReadableMap params, final Promise promise) {
        signalClient.saveSentMessage(tag, params.getString("username"), params.getString("message"), params.getInt("timestamp"), promise);
    }

    @ReactMethod
    public void decryptReceivedBody(String receivedBody, final Promise promise) {
        try {
            byte[] ciphertext = Base64.decode(receivedBody);
            LegacyMessage legacyMessage = new LegacyMessage(ciphertext, protocolStorage.getSignalingKey());
            byte[] messageBytes = legacyMessage.getBytes();
            promise.resolve(Base64.encodeBytes(messageBytes));
        } catch (IOException e) {
            e.printStackTrace();
            logSender.reportError(e);
            promise.reject(ERR_NATIVE_FAILED, e.getMessage());
        }
    }

    @ReactMethod
    public void decryptSignalMessage(String tag, String receivedMessage, final Promise promise) {
        signalClient.decryptSignalMessage(tag, receivedMessage, promise);
    }

    @ReactMethod
    public void deleteSignalMessage(String username, int timestamp, final Promise promise) {
        signalClient.deleteSignalMessage(username, timestamp, promise);
    }
}

File: ios/SignalMessagingSwift/Signal/RNSignalClientModule.swift
//
//  RNSignalClientModule.swift
//  myAwesomeProject
//
//  Created by Mantas on 07/07/2018.
//  Copyright © 2018 Facebook. All rights reserved.
//

import UIKit
import SignalProtocol
import Sentry

let ERR_WRONG_CONFIG = "ERR_WRONG_CONFIG"
let ERR_SERVER_FAILED = "ERR_SERVER_FAILED"
let ERR_NATIVE_FAILED = "ERR_NATIVE_FAILED"
let ERR_ADD_CONTACT_FAILED = "ERR_ADD_CONTACT_FAILED"

@objc(RNSignalClientModule)
class RNSignalClientModule: NSObject {

    private var signalResetVersion: Int
    private var username: String
    private var host: String
    private var signalClient: SignalClient

    override init() {
        self.signalResetVersion = 0
        self.username = ""
        self.host = ""
        self.signalClient = SignalClient(username: "", accessToken: "", host: "", logger: Logger(isLoggable: false), signalResetVersion: self.signalResetVersion)
        super.init()
    }

    @objc static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc func createClient(_ config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.username = config.object(forKey: "username") as! String
        self.host = config.object(forKey: "host") as! String
        let accessToken = config.object(forKey: "accessToken") as! String
        let isLoggable = config.object(forKey: "isSendingLogs") as! Bool
        let logger = Logger(isLoggable: isLoggable)
        let signalResetVersion = self.signalResetVersion
        let existingSignalResetVersion = ProtocolStorage().getSignalResetVersion()

        // ---
        // check pre key amount in order to reset from a version which was generating large amounts of prekeys
        let preKeys = ProtocolStorage().get(for: .PRE_KEYS_JSON_FILENAME)
        let performSoftReset = preKeys.count > 300
        // ---

        self.signalClient = SignalClient(username: username, accessToken: accessToken, host: host, logger: logger, signalResetVersion: signalResetVersion)
        if isLoggable {
            do {
                let sentryDSN = config.object(forKey: "errorTrackingDSN") as! String
                Client.shared = try Client(dsn: sentryDSN)
                try Client.shared?.startCrashHandler()
            } catch {
                // sentry unavailable
            }
        }

        if !performSoftReset && ProtocolStorage().getLocalUsername() == username && ProtocolStorage().isLocalRegistered() {
            self.signalClient.checkPreKeys()
            resolve("ok")
        } else {
            if (performSoftReset) {
                logger.sendInfoMessage(message: "iOS Signal soft reset version triggered: \(signalResetVersion)" )
                ProtocolStorage().destroyAllExceptMessages()
            } else {
                ProtocolStorage().destroyAll()
            }
            self.signalClient.register(success: { (success) in
                resolve(success)
            }) { (error, message) in
                reject(error, message, nil)
            }
        }
    }

    @objc func registerAccount(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.register(success: { (success) in
            resolve(success)
        }) { (error, message) in
            reject(error, message, nil)
        }
    }

    @objc func resetAccount(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        ProtocolStorage().destroyAll()
        resolve("ok")
    }

    @objc func addContact(_ config: NSDictionary, forceAdd: Bool, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let username = (config.object(forKey: "username") as? String) ?? ""
        let address = SignalAddress(name: username, deviceId: 1)
        if let result = self.signalClient.store()?.sessionStore.containsSession(for: address), result == false || forceAdd {
            self.signalClient.requestPreKeys(username: username, success: { (success) in
                resolve(success)
            }) { (error, message) in
                reject(error, message, nil)
            }
        } else {
            resolve("ok")
        }
    }

    @objc func setFcmId(_ fcmId: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.saveFCMId(fcmId: fcmId, success: {
            resolve("ok")
        }) { (error, message) in
            reject(error, message, nil)
        }
    }

    @objc func deleteContact(_ username: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let address = SignalAddress(name: username, deviceId: 1);
        _ = self.signalClient.store()?.sessionStore.deleteSession(for: address)
        ProtocolStorage().removeRemoteIdentity(for: address)
        MessagesStorage().deleteAllContactMessages(for: username)
        self.signalClient.deleteContactPendingMessages(username: username, messageTag: "*", success: { (success) in
            print("ok: pending messages deleted")
        }) { (error, message) in
            print("\(error): \(message)")
        }
        resolve("ok")
    }

    @objc func deleteContactMessages(_ username: String, messageTag: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        MessagesStorage().deleteContactMessages(for: username, tag: messageTag)
        self.signalClient.deleteContactPendingMessages(username: username, messageTag: messageTag, success: { (success) in
            resolve(success)
        }) { (error, message) in
            reject(error, message, nil)
        }
    }

    @objc func receiveNewMessagesByContact(_ username: String, messageTag: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.getContactMessages(username: username, messageTag: messageTag, decodeAndSave: true, success: { (success) in
            resolve(success)
        }) { (error, message) in
            reject(error, message, nil)
        }
    }

    @objc func getMessagesByContact(_ username: String, messageTag: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let sortedMessages = MessagesStorage().getMessages(for: username, tag: messageTag).sorted { (messageOne, messageTwo) -> Bool in
            return messageOne.savedTimestamp > messageTwo.savedTimestamp
            }.compactMap { (message) -> [String : Any]? in
                return message.dictionary
        }

        if let data = try? JSONSerialization.data(withJSONObject: sortedMessages, options: .prettyPrinted) {
            let jsonSring = String(data: data, encoding: .utf8)
            resolve(jsonSring)
        }
    }

    @objc func getExistingMessages(_ messageTag: String, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        if let data = try? JSONSerialization.data(withJSONObject: self.signalClient.getAllContactMessages(messageTag: messageTag), options: .prettyPrinted) {
            let jsonSring = String(data: data, encoding: .utf8)
            resolve(jsonSring)
        }
    }

    @objc func getUnreadMessagesCount(_ messageTag: String, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.getContactMessages(username: "", messageTag: messageTag, decodeAndSave: false, success: { (dict) in
            if let data = try? JSONSerialization.data(withJSONObject: dict, options: .prettyPrinted) {
                let jsonSring = String(data: data, encoding: .utf8)
                resolve(jsonSring)
            } else {
                reject("error", "serialization error in getUnreadMessagesCount", nil)
            }
        }) { (error, message) in
            reject(error, message, nil)
        }
    }

    @objc func sendMessageByContact(_ messageTag: String, config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.sendMessage(
            username: config.object(forKey: "username") as! String,
            messageString: config.object(forKey: "message") as! String,
            messageTag: messageTag,
            silent: false,
            success: { (success) in resolve(success) }
        ) { (error, message) in
            reject(error, message, nil)
        }
    }

    @objc func sendSilentMessageByContact(_ messageTag: String, config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.sendMessage(
            username: config.object(forKey: "username") as! String,
            messageString: config.object(forKey: "message") as! String,
            messageTag: messageTag,
            silent: true,
            success: { (success) in resolve(success) }
        ) { (error, message) in
            reject(error, message, nil)
        }
    }

    @objc func prepareApiBody(_ messageTag: String, config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let params = self.signalClient.prepareApiBody(
            username: config.object(forKey: "username") as! String,
            messageString: config.object(forKey: "message") as! String,
            messageTag: messageTag,
            silent: false,
            failure: { (error) in
                reject(ERR_NATIVE_FAILED, "\(error)", nil)
        }
        );

        if (params.isEmpty) {
            return
        }

        if let data = try? JSONSerialization.data(withJSONObject: params, options: .prettyPrinted) {
            let jsonSring = String(data: data, encoding: .utf8)
            resolve(jsonSring)
        }

    }

    @objc func saveSentMessage(_ messageTag: String, config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.saveSentMessage(
            messageTag: messageTag,
            username: config.object(forKey: "username") as! String,
            messageString: config.object(forKey: "message") as! String,
            timestamp: Int64(config.object(forKey: "timestamp") as! NSNumber)
        );
        resolve("ok")
    }

    @objc func decryptReceivedBody(_ receivedBody: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let decryptedBytes = self.signalClient.decryptReceivedBody(body: receivedBody)
        if (decryptedBytes.length == 0) {
            reject(ERR_NATIVE_FAILED, "Failed to decrypt received body", nil)
            return;
        }
        let encoded = decryptedBytes.base64EncodedString(options: NSData.Base64EncodingOptions.endLineWithLineFeed);
        if (encoded.count == 0) {
            reject(ERR_NATIVE_FAILED, "Failed to Base64 encode decrypted body", nil)
            return;
        }
        resolve(encoded)
    }

    @objc func decryptSignalMessage(_ messageTag: String, receivedMessage: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.decryptSignalMessage(
            messageTag: messageTag,
            receivedMessage: receivedMessage,
            success: { (message) in resolve(message) },
            failure: { (error) in
                reject(ERR_NATIVE_FAILED, "\(error)", nil)
        }
        )
    }


    @objc func deleteSignalMessage(_ username: String, timestamp: NSInteger, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        self.signalClient.deleteSignalMessage(
            username: username,
            timestamp: timestamp,
            success: { (message) in resolve(message) }
        )
    }
}
